platform/darwin/src/MGLGeoJSONSource.mm

```diff
@@ -118,11 +127,28 @@ - (void)validateValue:(id)value
     if (self.URL) {
         NSURL *url = self.URL.mgl_URLByStandardizingScheme;
         source->setURL(url.absoluteString.UTF8String);
-    } else {
+    } else if (self.geoJSONData) {
         NSString *string = [[NSString alloc] initWithData:self.geoJSONData encoding:NSUTF8StringEncoding];
         const auto geojson = mapbox::geojson::parse(string.UTF8String).get<mapbox::geojson::feature_collection>();
         source->setGeoJSON(geojson);
         _features = MGLFeaturesFromMBGLFeatures(geojson);
+    } else {
+    
+        NSMutableArray<NS_DICTIONARY_OF(NSString *, id) *> *featuresArray = [NSMutableArray array];
+        for (id<MGLFeature> feature in self.features) {
+            [featuresArray addObject:[feature featureDictionary]];
+        }
+        
+        NS_DICTIONARY_OF(NSString *, id) *featureCollection = @{@"type":@"FeatureCollection",
+                                            @"features":featuresArray};
+        
+        NSError *error;
+        NSData *featuresJSONData = [NSJSONSerialization dataWithJSONObject:featureCollection options:NSJSONWritingPrettyPrinted error:&error];
```

Comment:
Pretty printing is good for debugging purposes, but it'll need to be removed before merging because it increases memory consumption.